### PR TITLE
hw_53

### PR DESCRIPTION
**Pull request by Vladyslav-Markevych**

### CODE REVIEW

File: express-server.js [hunk 1]
```diff
@@ -3,11 +3,49 @@ import "dotenv/config";
 import bodyParser from "body-parser";
 import { users, products, carts, orders } from "./storage.js";
 import crypto from "crypto";
-import { CustomError, ValidationError, Unauthorized, NotFound } from "./errorHandler.js";
+import fs from "fs";
+import path, { dirname } from "path";
```
Comment (VitalyaIlin) [inline express-server.js:7]: why do you import `dirname` here?

File: express-server.js [hunk 2]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
   }
 });
 
+// hw_53
+app.post("/product", async (req, res) => {
+  const userId = checkUserById(req.headers["x-user-id"]);
```
Comment (VitalyaIlin) [inline express-server.js:228]: Why do you need this line of code?

File: express-server.js [hunk 3]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+
+  let productJSON = await readJsonFile(filePathJson);
+
+  productJSON.push(obj);
+
+  fs.writeFile(
+    filePathJson,
+    JSON.stringify(productJSON, null, 2),
+    "utf8",
+    (err) => {
```
Comment (VitalyaIlin) [inline express-server.js:253]: Handle this error

File: express-server.js [hunk 4]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+    await fs.promises.mkdir(directoryPath, { recursive: true });
+
+    const filePath = path.join(directoryPath, `${id}.jpg`);
+    eventEmitter.emit("fileUploadStart");
+
+    let productJSON = await readJsonFile(filePathJson);
+    const obj = productJSON.find((item) => item.id == productId);
+
+    if (obj) {
+      obj.image.push(`${id}.jpg`);
```
Comment (VitalyaIlin) [inline express-server.js:275]: Create a new variable for the file name

File: express-server.js [hunk 5]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+
+    const filePath = path.join(directoryPath, `${id}.mp4`);
+    eventEmitter.emit("fileUploadStart");
+
+    const filePathJson = path.join(__dirname, "./products.store.json");
+    let productJSON = await readJsonFile(filePathJson);
+    const obj = productJSON.find((item) => item.id == productId);
+
+    if (obj) {
+      obj.video.push(`${id}.mp4`);
```
Comment (VitalyaIlin) [inline express-server.js:327]: create a new variable for the file name

File: express-server.js [hunk 6]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+  } catch (err) {
+    eventEmitter.emit("fileUploadFailed", err);
+    res.status(500).send("Server error");
+  }
+});
+app.get("/product/video/:fileName", async (req, res) => {
+  const { fileName } = req.params;
+  const filePathJson = path.join(__dirname, "./products.store.json");
+  try {
+    eventEmitter.emit("fileUploadStart");
```
Comment (VitalyaIlin) [inline express-server.js:407]: It's not necessary to emit such events here they are only for file upload

File: express-server.js [hunk 7]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+    const videoPath = path.join(
+      __dirname,
+      "productFiles",
+      obj.id,
+      "video",
+      NameFile
+    );
+    const stream = fs.createReadStream(videoPath);
+    stream.on("open", () => {
+      eventEmitter.emit("fileUploadEnd");
```
Comment (VitalyaIlin) [inline express-server.js:423]: it's not necessary

File: express-server.js [hunk 8]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+    });
+
+    stream.on("error", (error) => {
+      eventEmitter.emit("fileUploadFailed", error);
```
Comment (VitalyaIlin) [inline express-server.js:427]: it's not necessary

File: express-server.js [hunk 9]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+      res.status(500).send("Server error while reading the file");
+    });
+
+    stream.pipe(res);
+  } catch (err) {
+    eventEmitter.emit("fileUploadFailed", err);
```
Comment (VitalyaIlin) [inline express-server.js:433]: it's not necessary

File: express-server.js [hunk 10]
```diff
@@ -165,6 +223,218 @@ app.post("/api/cart/checkout", isAuth, (req, res) => {
...
+    res.status(500).send("Server error");
+  }
+});
+app.get("/product/video/:fileName", async (req, res) => {
+  const { fileName } = req.params;
+  const filePathJson = path.join(__dirname, "./products.store.json");
+  try {
+    eventEmitter.emit("fileUploadStart");
+    let productJSON = await readJsonFile(filePathJson);
+    const NameFile = fileName + ".mp4";
```
Comment (VitalyaIlin) [inline express-server.js:409]: rename this variable. Example: `fullFileName`

**Review by VitalyaIlin — COMMENTED** [review]
review the comments

File: express-server.js (hunk 1 above)
Comment (Vladyslav-Markevych) [reply express-server.js:7]: Автоматически подтянул VScode

**Pull request merged**